powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/repository/ActivationRepository.java

```diff
@@ -156,4 +156,15 @@ default Long getActivationCountByActivationCode(Long applicationId, String activ
     @Query("SELECT a FROM ActivationRecordEntity a WHERE a.application.id = ?1 AND a.activationCode LIKE ?2% AND a.activationStatus IN ?3 AND a.timestampActivationExpire > ?4")
     ActivationRecordEntity findCreatedActivationByShortIdWithoutLock(Long applicationId, String activationIdShort, Collection<ActivationStatus> states, Date currentTimestamp);
 
+    /**
+     * Find all activations which match the query criteria.
+     * @param userIds List of user IDs, at least one user ID should be specified.
+     * @param applicationIds List of application IDs, use null value for all applications.
+     * @param timestampLastUsedBefore Last used timestamp (timestampLastUsed < timestampLastUsedBefore), use null value for any date.
+     * @param timestampLastUsedAfter Last used timestamp (timestampLastUsed >= timestampLastUsedAfter), use null value for any date.
+     * @param states List of activation states to consider.
+     * @return List of activations which match the query criteria.
+     */
+    @Query("SELECT a FROM ActivationRecordEntity a WHERE a.userId IN ?1 AND (?2 IS NULL OR a.application.id IN ?2) AND (?3 IS NULL OR a.timestampLastUsed < ?3) AND (?4 IS NULL OR a.timestampLastUsed >= ?4) AND a.activationStatus IN ?5")
```

Comment:
Should we use named parameters for better readability?

https://www.baeldung.com/spring-data-jpa-query#named-parameters

Reply:
Yes, we can.